runtime/j9vm/javanextvmi.cpp

```diff
@@ -274,21 +274,57 @@ enterVThreadTransitionCritical(J9VMThread *currentThread, jobject thread)
 	MM_ObjectAccessBarrierAPI objectAccessBarrier = MM_ObjectAccessBarrierAPI(currentThread);
 	j9object_t threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
 
-	while(!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, threadObj, vm->virtualThreadInspectorCountOffset, 0, (U_64)-1)) {
+retry:
+	while (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, threadObj, vm->virtualThreadInspectorCountOffset, 0, (U_64)-1)) {
 		/* Thread is being inspected or unmounted, wait. */
 		vmFuncs->internalReleaseVMAccess(currentThread);
 		VM_AtomicSupport::yieldCPU();
 		/* After wait, the thread may suspend here. */
 		vmFuncs->internalAcquireVMAccess(currentThread);
 		threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
 	}
+
+	/* Link the current J9VMThread with the virtual thread object. */
+	if (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, threadObj, vm->isSuspendedInternalOffset, 0, (U_64)currentThread)) {
+		/* If virtual thread is suspended while unmounted, reset the inspectorCount and do a wait and retry. */
+		if (J9_ARE_ANY_BITS_SET(J9OBJECT_U64_LOAD(currentThread, threadObj, vm->isSuspendedInternalOffset), 1)) {
+			J9OBJECT_I64_STORE(currentThread, threadObj, vm->virtualThreadInspectorCountOffset, 0);
+		}
+		vmFuncs->internalReleaseVMAccess(currentThread);
+		/* Spin is used instead of the halt flag as we cannot guarantee suspend flag is still set now.
+		 *
+		 * TODO: Dynamically increase the sleep time to a bounded maximum.
+		 */
+		f_threadSleep(10);
+		/* After wait, the thread may suspend here. */
+		vmFuncs->internalAcquireVMAccess(currentThread);
+		threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
+		goto retry;
+	}
 }
 
 static void
-exitVThreadTransitionCritical(J9VMThread *currentThread, j9object_t vthread)
+exitVThreadTransitionCritical(J9VMThread *currentThread, jobject thread)
 {
-	Assert_SC_true(-1 == J9OBJECT_I64_LOAD(currentThread, vthread, currentThread->javaVM->virtualThreadInspectorCountOffset));
-	J9OBJECT_I64_STORE(currentThread, vthread, currentThread->javaVM->virtualThreadInspectorCountOffset, 0);
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	j9object_t vthread = J9_JNI_UNWRAP_REFERENCE(thread);
+	MM_ObjectAccessBarrierAPI objectAccessBarrier = MM_ObjectAccessBarrierAPI(currentThread);
+	U_64 isSuspendedFlag = J9OBJECT_U64_LOAD(currentThread, vthread, vm->isSuspendedInternalOffset);
+
+	/* Remove J9VMThread address from isSuspendedInternal field, as the thread state is no longer in a transition. */
+	while (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, vthread, vm->isSuspendedInternalOffset, (U_64)currentThread, 0)) {
+		/* Wait if the suspend flag is set. */
+		vmFuncs->internalReleaseVMAccess(currentThread);
+		VM_AtomicSupport::yieldCPU();
```

Comment:
similar to the `enterVThreadTransitionCritical`, we should use here `f_threadSleep` since it is more granular and it will reduce CPU hogging.